core/templates/vset.h

```diff
@@ -80,12 +80,12 @@ class VSet {
 			return -1;
 		}
 
-		int low = 0;
+		uint32_t low = 0;
 		int high = _data.size() - 1;
-		int middle;
+		uint32_t middle;
 		const T *a = &_data[0];
 
-		while (low <= high) {
+		while ((int)low <= high) {
```

Comment:
Wouldn't it be possible to streamline the types a bit more?
Seems weird to me to use `uint32_t` for only `low` and `middle` but then expose ourselves to truncation in `middle = (low + high) / 2` if `high` is big.

Reply:
`while ((int)low <= high)` here `high` can be `-1` and `low` `0` and we should exit the loop.
 `middle = (low + high) / 2` here, the compiler does this `middle = ((low + static_cast<unsigned int>(high)) / 2);`

Reply:
I was thinking of overflow if `high` is bigger than `UINT32_MAX`.

Reply:
The compiler with optimization enabled fully uses `high` as a `uint` and adds another check here.
``` c++
if (p_val < a[middle].key) {
    high = middle - 1; //search low end of array
    if(high == 0)
    {
        break;
    }
}
```
```
lea     edx, [rax-1]
cmp     ecx, edx
jg      .L109
```

Reply:
> I was thinking of overflow if `high` is bigger than `UINT32_MAX`.

This structure cannot be larger than int32_max, since we return int